### ISSUE TITLE
FAB - Nothing happens when Start chat is clicked after creating a room offline

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#40855]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.65-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Found when executing:** https://github.com/Expensify/Expensify/issues/302054
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
1. Log in with a new account
2. Create a workspace
3. Go offline
4. Create a new room
5. Navigate to the LHN
6. Navigate to FAB - Start chat

## Expected Result:
Start chat modal should open.

## Actual Result:
Nothing happens when Start chat is clicked after creating a room offline. On iOS and Android native, you can see it opening for a short time but it disappears after. On mWeb Safari, the page shakes after the tap.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/e4b393d6-5868-489c-ad9d-6e7e5ea296a4

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When we open start chat again after creating a room while offline, nothing happen.

### What is the root cause of that problem?
When we create a new room, it will optimistically set the loading to the new room form
https://github.com/Expensify/App/blob/08b7ff41656a34373761fb2bd505cc0dfa64be7b/src/libs/actions/Report.ts#L1955-L1957

Because we are offline, the loading state is always true. When we reopen the page again, it will immediately close because of the below effect.
https://github.com/Expensify/App/blob/08b7ff41656a34373761fb2bd505cc0dfa64be7b/src/pages/workspace/WorkspaceNewRoomPage.tsx#L146-L152

The `(isOffline && isLoading)` condition is true. When the page is mounted, we clear the loading state (set to false)
https://github.com/Expensify/App/blob/08b7ff41656a34373761fb2bd505cc0dfa64be7b/src/pages/workspace/WorkspaceNewRoomPage.tsx#L128-L130

and this condition `(wasLoading && !isLoading)` is also becomes true.

### What changes do you think we should make in order to solve the problem?

The `useEffect` is added (in https://github.com/Expensify/App/pull/31655) so it will only close the page when the request is successful, but we added back the code to close it immediately in `Report.addPolicyReport`, so it's really weird, I think it's a bad merge, so we can remove the dismiss code here.
https://github.com/Expensify/App/blob/08b7ff41656a34373761fb2bd505cc0dfa64be7b/src/libs/actions/Report.ts#L2012-L2013

To fix this issue, we can set `initWithStoredValues: false` to the new room form state. This way, the data will always be empty initially. We don't want the previous state to affect the new page.
```
formState: {
    key: ONYXKEYS.FORMS.NEW_ROOM_FORM,
    initWithStoredValues: false,
},
```

Then, we can remove this effect
https://github.com/Expensify/App/blob/08b7ff41656a34373761fb2bd505cc0dfa64be7b/src/pages/workspace/WorkspaceNewRoomPage.tsx#L128-L130

Or trigger it when it's unmounted instead.

NOTE: there is currently a bug in Onyx where `initWithStoredValues` doesn't work. To fix it we should initialize the onyx key only if `initWithStoredValues` is true.

```diff
-if (
+if (mapping.initWithStoredValues &&
    ((value !== undefined
        && !Onyx.hasPendingMergeForKey(key))
    || mapping.allowStaleData)
) {
```

**Comment by muttmuure** [comment]
@mollfpr there's a proposal above for you to take a look at

**Comment by mollfpr** [comment]
@bernhardoj Is there any conversation regarding the bug of `initWithStoredValues`? It should be critical if the property doesn't work.

**Comment by bernhardoj** [comment]
@mollfpr I'm not aware of any conversation of that

**Comment by mollfpr** [comment]
The @bernhardoj [proposal](https://github.com/Expensify/App/issues/40855#issuecomment-2074862188) looks good to me! For the issue on `initWithStoredValues`, I think we can decide after the PR is created.

🎀 👀 🎀 C+ reviewed!

**Comment by puneetlath** [comment]
>For the issue on initWithStoredValues, I think we can decide after the PR is created.

Sorry, I don't follow what this means. Mind explaining for me?

**Comment by mollfpr** [comment]
I notice that `initWithStoredValues` is only works for the first time the page rendered.


https://github.com/Expensify/App/assets/25520267/15d30dbe-7296-4ee0-b70a-c875732004cd

Here's in the video, after the refresh and then opening the new room page, you can see in the console log it will show `undefined` which is expected because I set `initWithStoredValues` to `false`. But the second time I opened the page, it didn't show `undefined` but instead the value stored.

I'm not sure if this is expected and what causing the issue, but @bernhardoj may be another option is to clear the `formState` after it unmounts can be the option to dodge the issue?

**Comment by bernhardoj** [comment]
> may be another option is to clear the formState after it unmounts can be the option to dodge the issue?

It won't work if the user closes the app/tab. We need to fix the onyx first before creating the App PR. We can create a local state of loading with an initial value of false, but I think it's better to fix the onyx instead of avoiding it.

**Comment by mollfpr** [comment]
@bernhardoj Could you create a discussion of the onyx issue? We can confirm if it is an issue before fixing it.

**Comment by bernhardoj** [comment]
I think the property name is self-explanatory and currently it doesn't work as the name and comment suggests, 
<img width="733" alt="image" src="https://github.com/Expensify/App/assets/50919443/c4ffc0cb-8ab8-4e47-a3b2-0b3762a22d41">

but here is the [discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1715059904152869).

**Comment by puneetlath** [comment]
Ok understood. So my apologies, can you summarize for me the changes we'll be making in Onyx and in App?

**Comment by bernhardoj** [comment]
@mollfpr so, based on the discussion, it's confirmed the current behavior of `initWithStoredValues` is wrong. To further prove that it's a bug, we can run a unit test. In onyx repo itself, there is already a unit test for withOnyx + `initWithStoredValues`, but the test case is currently wrong.

<img width="650" alt="image" src="https://github.com/Expensify/App/assets/50919443/c3ba7d20-7d28-49c3-b7f3-75141bd7cfe6">

You can see the key name that is set to Onyx ('test_key') and is connected to the component ('test') is different. That's why the test will always pass. You can update the 'test_key' to 'test' and see it will fails.

@puneetlath the plan will be,
1. Update the onyx unit test case
2. Apply the onyx `initWithStoredValues` bug fix in [here](https://github.com/Expensify/react-native-onyx/blob/d7ab22617fedc1faba251e9680909793cd01ea9c/lib/withOnyx.js#L78-L81)
<img width="878" alt="image" src="https://github.com/Expensify/App/assets/50919443/699c753a-e252-4d5e-b735-8b39dbb13165">
3. Update the onyx in the App and apply the App fix
<img width="868" alt="image" src="https://github.com/Expensify/App/assets/50919443/ec639548-01d4-403b-80d9-b9c289be3c86">

_the same issue actually happens in useOnyx hook too, but I'm not confident to fix the new hook. `initWithStoredValues` currently only works with Onyx.connect_

**Comment by mollfpr** [comment]
@bernhardoj The plan looks good to me!

**Comment by puneetlath** [comment]
Thanks for laying that all out @bernhardoj. Sounds great.

**Comment by bernhardoj** [comment]
The Onyx PR is [here](https://github.com/Expensify/react-native-onyx/pull/545)

cc: @mollfpr @puneetlath

**Comment by bernhardoj** [comment]
Onyx PR is merged. The current onyx version is 2.0.32 but we can't update it yet because [2.0.33 version has a regression](https://github.com/Expensify/react-native-onyx/pull/519#issuecomment-2097136863), so we need to wait until the regression is fixed.

**Comment by mollfpr** [comment]
@bernhardoj We have this bump https://github.com/Expensify/App/pull/42057, are the changes included in the `2.0.40`?

**Comment by bernhardoj** [comment]
Our Onyx PR is in 2.0.39, so yes it will be included in 2.0.40. We can hold for that then.

**Comment by puneetlath** [comment]
Do we have an issue for the Onyx bump?

**Comment by puneetlath** [comment]
I think Onyx has been bumped past this version. What's left to do now?

**Comment by bernhardoj** [comment]
What's left is the App PR. Here is the App [PR](https://github.com/Expensify/App/pull/42679).

cc: @mollfpr

**Comment by puneetlath** [comment]
@mollfpr bump on the checklist so we can pay.

**Comment by bernhardoj** [comment]
@puneetlath I think we still need to wait a bit based on my comment [here](https://github.com/Expensify/App/pull/42679#issuecomment-2143209025)

**Comment by bernhardoj** [comment]
@puneetlath The Onyx PR is merged. I think we can have QA retest this.

**Comment by kavimuru** [comment]
Bug fixed and not reproducible 

https://github.com/Expensify/App/assets/43996225/6fc4ce3e-f106-4fa8-83b5-1168f9bca6bf

**Comment by puneetlath** [comment]
Ok, so we're ready for the checklist now yeah @bernhardoj @mollfpr? So that we can pay?

**Comment by bernhardoj** [comment]
Yes @mollfpr

**Comment by mollfpr** [comment]
>[@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
 [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

No offending PR.

> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
 [@mollfpr] Determine if we should create a regression test for this bug.

The regression step should be enough.

> [@mollfpr] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

1. Go offline
2. Create a new room from FAB
3. Open the new room page again from FAB
4. Verify the new room page is shown
5. 👍 or 👎

**Comment by muttmuure** [comment]
@mollfpr - $250 C+

**Comment by JmillsExpensify** [comment]
$250 approved for @mollfpr

**Comment by puneetlath** [comment]
@muttmuure what is left to do here? Can we close?

**Comment by bernhardoj** [comment]
I think the payment for me is still pending.

**Comment by muttmuure** [comment]
@bernhardoj paid $250 for C
